mbs_results/pivot_imputation_value.py

```diff
@@ -0,0 +1,169 @@
+
+    Parameters
+    ----------
+    dataframe : pd.DataFrame
+        Reference dataframe containing links, count values, and imputed values
+        by identifier, cell, date, and question
+    identifier : str
+        name of column in dataframe containing identifier variable
+    date : str
+        name of column in dataframe containing date variable
+    sic : str
+        name of column in dataframe containing sic variable
+    cell : str
+        name of column in dataframe containing cell variable
+    forward : str
+        name of column in dataframe containing forward link variable
+    backward : str
+        name of column in dataframe containing backward link variable
+    construction : str
+        name of column in dataframe containing construction link variable
+    question : str
+        name of column in dataframe containing question code variable
+    imputed_value: str
+        name of column in dataframe containing imputed_value variable
+    f_count: str,
+        name of column in dataframe containing f_count variable
+    b_count: str,
+        name of column in dataframe containing b_count variable
+    c_count: str,
+        name of column in dataframe containing c_count variable
+    selected_periods: list,
+        list containing periods to include in output
+
+    Returns
+    -------
+    dataframe filtered by date, containing imputation_value, pivoted by imputation type
+    and grouped by sic, cell, question and imputation type.
+
+    """
+    if selected_periods is not None:
+        df = df.query("{} in {}".format(date, selected_periods))
+
+    links_df = df.melt(
+        id_vars=[date, sic, cell, question, imputed_value],
+        value_vars=[forward, backward, construction],
+        var_name="link_type",
+        value_name="imputation_link",
+    )
+
+    link_type_map = {forward: "F", backward: "B", construction: "C"}
+    links_df["link_type"] = links_df["link_type"].map(link_type_map)
+
+    counts_df = df.melt(
+        id_vars=[date, sic, cell, question],
+        value_vars=[f_count, b_count, c_count],
+        var_name="link_type_count",
+        value_name="count",
+    )
+
+    link_type_map_count = {f_count: "F", b_count: "B", c_count: "C"}
+    counts_df["link_type_count"] = counts_df["link_type_count"].map(link_type_map_count)
+
+    merged_df = pd.merge(
+        links_df,
+        counts_df,
+        how="outer",
+        left_on=[date, sic, cell, question, "link_type"],
+        right_on=[date, sic, cell, question, "link_type_count"],
+    )
+
+    merged_df.drop_duplicates(inplace=True)
+    merged_df.drop(["link_type_count"], axis=1, inplace=True)
+
+    merged_df = merged_df.groupby(
+        [date, sic, cell, question, "link_type"], as_index=False
+    ).agg({imputed_value: "sum", "count": "first", "imputation_link": "first"})
+
+    sorting_order = {"F": 1, "B": 2, "C": 3}
+    merged_df["sort_column"] = merged_df["link_type"].map(sorting_order)
```

Comment:
Is sorting a requirement ? Nice solution, since link_type order matters we could convert that column to categorical and specify the order https://pandas.pydata.org/pandas-docs/version/1.1/reference/api/pandas.Categorical.html

Reply:
Do you mean something like:

merged_df['link_type'] = pd.Categorical(merged_df['link_type'], categories=['F','B','C'], ordered=True)
merged_df.sort_values(by=['link_type'], axis=0, ascending=True, inplace=True)

I'm struggling to get this (or similar) to work and pass Pytest but I could just be missing something obvious. The ordering of this column was the hardest part to get right in the first place - every time I made a change to the code it seemed to randomly change the order of link_type!

I've reverted back to the original for the time being but happy to discuss.